{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "<script>\n",
    "  jQuery(document).ready(function($) {\n",
    "\n",
    "  $(window).load(function(){\n",
    "    $('#preloader').fadeOut('slow',function(){$(this).remove();});\n",
    "  });\n",
    "\n",
    "  });\n",
    "</script>\n",
    "\n",
    "<style type=\"text/css\">\n",
    "  div#preloader { position: fixed;\n",
    "      left: 0;\n",
    "      top: 0;\n",
    "      z-index: 999;\n",
    "      width: 100%;\n",
    "      height: 100%;\n",
    "      overflow: visible;\n",
    "      background: #fff url('http://preloaders.net/preloaders/720/Moving%20line.gif') no-repeat center center;\n",
    "  }\n",
    "\n",
    "</style>\n",
    "\n",
    "<div id=\"preloader\"></div>"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "<script>\n",
    "  function code_toggle() {\n",
    "    if (code_shown){\n",
    "      $('div.input').hide('500');\n",
    "      $('#toggleButton').val('Show Code')\n",
    "    } else {\n",
    "      $('div.input').show('500');\n",
    "      $('#toggleButton').val('Hide Code')\n",
    "    }\n",
    "    code_shown = !code_shown\n",
    "  }\n",
    "\n",
    "  $( document ).ready(function(){\n",
    "    code_shown=false;\n",
    "    $('div.input').hide()\n",
    "  });\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" id=\"toggleButton\" value=\"Show Code\"></form>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load rpy2 magic\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "# to switch off warning messages\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# make default cell width 85% of available screen\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:85% !important; }</style>\"))\n",
    "\n",
    "# load R libraries & functions\n",
    "%R options(warn=-1)\n",
    "%R library(RColorBrewer)\n",
    "%R library(ComplexHeatmap)\n",
    "%R library(circlize)\n",
    "%R library(dendextend)\n",
    "%R library(Rtsne)\n",
    "%R library(ggplot2)\n",
    "%R library(gplots)\n",
    "%R library(gridExtra)\n",
    "%R library(DESeq2)\n",
    "%R library(scales)\n",
    "%R library(dplyr)\n",
    "%R library(dendextend)\n",
    "%R library(ggrepel)\n",
    "%R library(wesanderson)\n",
    "%R library(vsn)\n",
    "%R library(reshape2)\n",
    "%R library(dplyr)\n",
    "%R library(plyr)\n",
    "%R library(topGO)\n",
    "%R library(stringr)\n",
    "\n",
    "# supress R warnings/messages\n",
    "%R options(warn=-1, echo=FALSE, message=FALSE, results='hide', verbose=FALSE)\n",
    "\n",
    "# load python modules\n",
    "import glob\n",
    "import re\n",
    "import sys\n",
    "import os\n",
    "import rpy2.robjects as robjects\n",
    "import sqlite3\n",
    "import yaml\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "db = \"./csvdb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get options pipeline.yml\n",
    "import yaml\n",
    "with open(\"pipeline.yml\") as o:\n",
    "    opts = yaml.load(o)\n",
    "#     print(opts)\n",
    "\n",
    "anndb = opts[\"annotations\"][\"database\"]\n",
    "# print(anndb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_DataFrame(query, dbhandle=\"csvdb\"):\n",
    "    '''Fetch query results and returns them as a pandas dataframe'''\n",
    "\n",
    "    dbhandle = sqlite3.connect(dbhandle)\n",
    "\n",
    "    cc = dbhandle.cursor()\n",
    "    sqlresult = cc.execute(query).fetchall()\n",
    "    cc.close()\n",
    "\n",
    "    # see http://pandas.pydata.org/pandas-docs/dev/generated/\n",
    "    # pandas.DataFrame.from_records.html#pandas.DataFrame.from_records\n",
    "    # this method is design to handle sql_records with proper type\n",
    "    # conversion\n",
    "\n",
    "    field_names = [d[0] for d in cc.description]\n",
    "    pandas_DataFrame = pd.DataFrame.from_records(\n",
    "        sqlresult,\n",
    "        columns=field_names\n",
    "    )\n",
    "    return pandas_DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# R functions\n",
    "theme_notebook <- function(base_size=18, base_family=\"helvetica\") {\n",
    "                  (theme_set(theme_minimal(base_size=18))\n",
    "                  + theme(plot.title = element_text(face=\"bold\", size=20, hjust=0.5),\n",
    "                             text = element_text(),\n",
    "                             axis.title = element_text(face=\"bold\",size = rel(1)),\n",
    "                             axis.title.y = element_text(angle=90,vjust=2, size=20),\n",
    "                             axis.title.x = element_text(vjust=-0.2, size=20),\n",
    "                             axis.text = element_text(size=20),\n",
    "                             axis.line = element_line(colour=\"black\"),\n",
    "                             axis.ticks = element_line(),\n",
    "                             legend.key = element_rect(colour = NA),\n",
    "                             legend.key.size= unit(0.5, \"cm\"),\n",
    "                             legend.margin = unit(0.5, \"cm\"),\n",
    "                             legend.text = element_text(size=14),\n",
    "                             legend.title = element_text(size=16),\n",
    "                             ))\n",
    "}\n",
    "\n",
    "ggplot_prcomp <- function(prcomp_object,\n",
    "                          plots=list(\"A\"=c(\"PC1\",\"PC2\"), \"B\"=c(\"PC3\",\"PC4\"), \"C\"=c(\"PC5\",\"PC6\")),\n",
    "                          sample_information=\"none\",\n",
    "                          color=\"c()\",\n",
    "                          shape=\"c()\",\n",
    "                          label=\"none\",\n",
    "                          size=3,\n",
    "                          nudge_scale_factor=40){\n",
    "    require(gridExtra)\n",
    "    pca = prcomp_object\n",
    "\n",
    "    # sample_information should have the same rownames as pca$x\n",
    "    pvs <- summary(pca)$importance[\"Proportion of Variance\",]\n",
    "\n",
    "    names = paste(names(pvs),\" (\",round(pvs,2),\")\",sep=\"\")\n",
    "\n",
    "    #scree plot\n",
    "    fs <- data.frame(x=c(1:length(names(pvs))), y=as.vector(pvs))\n",
    "\n",
    "    pcdf <- as.data.frame(pca$x)\n",
    "\n",
    "    pcdf <- merge(pcdf, sample_information, by=0, all=T)\n",
    "\n",
    "    gps = list()\n",
    "\n",
    "    scree <- ggplot(fs, aes(x,y)) + \n",
    "                geom_point(size=4) + \n",
    "                xlab(\"principal component\") + \n",
    "                ylab(\"proportion of variance\") + \n",
    "                ggtitle(\"scree plot\") +\n",
    "                theme_notebook()\n",
    "\n",
    "    c_lab <- function(props, C){\n",
    "        return(paste(C, \" (\", props[[C]]*100,\"%)\",sep=\"\"))\n",
    "    }\n",
    "\n",
    "    for(plot in names(plots)){\n",
    "\n",
    "        comps <- plots[[plot]]\n",
    "\n",
    "        PCX <- comps[1]\n",
    "        PCY <- comps[2]\n",
    "\n",
    "        nudge_x <- diff(range(pcdf[[PCX]]))/nudge_scale_factor\n",
    "        nudge_y <- diff(range(pcdf[[PCY]]))/nudge_scale_factor\n",
    "\n",
    "        gp <- ggplot(pcdf, aes_string(PCX, PCY, color=color, shape=shape)) + theme_notebook()\n",
    "\n",
    "        if(label!=\"none\"){\n",
    "            gp <- gp + geom_text(aes_string(label=label), nudge_x=nudge_x, nudge_y=nudge_y, color=\"black\")\n",
    "        }\n",
    "\n",
    "        gp <- gp + geom_point(size=size)\n",
    "        gp <- gp + xlab(c_lab(pvs,PCX)) + ylab(c_lab(pvs,PCY))\n",
    "\n",
    "        gps[[plot]] <- gp\n",
    "\n",
    "    }\n",
    "\n",
    "    gps[[\"scree\"]] <- scree\n",
    "\n",
    "    return(gps)\n",
    "}\n",
    "\n",
    "plot_heatmap <- function(df, sampleinfo, palette, clust_method=\"ward.D2\", dist_method=\"manhattan\", low_opt=0.1, \n",
    "                         med_opt=FALSE, high_opt=0.9, labels=\"random\", no_labels=50, scale_data=\"other\", clust=0, \n",
    "                         order=NULL, scale_colours=c(\"royalblue3\", \"white\", \"red3\"), row_clust=TRUE, col_clust=TRUE,\n",
    "                         title=\"\", legend_label=FALSE){\n",
    "\n",
    "     ### complex heatmap plot ###\n",
    "    # input: data frame, output: plots heatmap and returns df ordered by dendogram\n",
    "        # gene_name must be a column in df!\n",
    "    # Optionally can split row dendogram in to up to 10 clusters, which are annotated on returned df\n",
    "    \n",
    "    require(wesanderson)\n",
    "    require(dendextend)\n",
    "    require(ComplexHeatmap)\n",
    "    require(circlize)\n",
    "    require(dplyr)\n",
    "    require(plyr)\n",
    "    \n",
    "    rownames(df) <- NULL # these causes problems with gene labels if not integers\n",
    "    \n",
    "    # get vector of labels\n",
    "    if (labels == \"random\"){\n",
    "        subset <- df[sample(no_labels, replace=FALSE), ]\n",
    "        subset <- as.integer(rownames(subset))\n",
    "        labels <- df$gene_name[subset]\n",
    "        show_rnames <- FALSE\n",
    "        lsize <- 12\n",
    "    } \n",
    "    \n",
    "    # label top variable genes\n",
    "    else if (labels == \"top_var\"){ # get most variable genes at day2\n",
    "        subset <- cbind(df)\n",
    "        genes <- subset$gene_name # remove labels\n",
    "        subset$gene_name <- NULL\n",
    "        lsize <- 12\n",
    "        subset$var_genes <- apply(subset, 1, sd)# get standard deviation of genes across samples\n",
    "        subset$gene_name <- genes\n",
    "        subset <- subset[order(subset$var_genes, decreasing=TRUE), ] # order by most variable genes\n",
    "        subset <- head(subset, no_labels) # limit to top var genes\n",
    "        subset <- as.integer(rownames(subset)) # get index\n",
    "        labels <- df[subset, ]$gene_name # subset df by index\n",
    "        show_rnames <- FALSE\n",
    "    } else if (labels==\"all\"){\n",
    "        show_rnames <- FALSE\n",
    "        lsize <- 8\n",
    "        subset <- cbind(df) # make a copy\n",
    "        subset <- as.integer(rownames(subset)) # get index\n",
    "        labels <- df[subset, ]$gene_name # subset df by index\n",
    "    }\n",
    "    else{\n",
    "        print(\"Specify what to label, options= random, top_var\")\n",
    "    }\n",
    "    \n",
    "    rnames <- df$gene_name\n",
    "    cnames <- rownames(sampleinfo)\n",
    "    df$gene_name <- NULL\n",
    "\n",
    "    matrix <- data.matrix(df)\n",
    "    matrix[is.na(matrix)] <- 0\n",
    "    \n",
    "    rownames(matrix) <- rnames\n",
    "    colnames(matrix) <- cnames\n",
    "\n",
    "    if (scale_data == \"min_max\"){\n",
    "        # scale each row (gene) manually to min - max of row\n",
    "        matrix <- t(apply(matrix, 1, function(x)(x-min(x))/(max(x)-min(x))))\n",
    "        legend_name <- \"Relative Expression\"\n",
    "    }\n",
    "    else if (scale_data == \"z_score\") {\n",
    "        # z-score each row of matrix\n",
    "        matrix <- t(scale(t(matrix)))\n",
    "        legend_name <- \"Z-Score\"\n",
    "    }\n",
    "    else {\n",
    "    #         print(\"No scaling\")\n",
    "        legend_name <- \"Expression\"\n",
    "    }\n",
    "\n",
    "    # Calculate distance\n",
    "    if(dist_method == \"minkowski\"){\n",
    "        p=1.44 # If method is \"minkowski\" set p as 1.44\n",
    "        dist_genes <- dist(matrix, method=dist_method, p)\n",
    "        }\n",
    "    else{\n",
    "        dist_genes <- dist(matrix, method=dist_method)\n",
    "        }\n",
    "\n",
    "    # Clustering\n",
    "    clust_genes <- hclust(dist_genes, method=clust_method)\n",
    "\n",
    "    if (clust){\n",
    "        # split dendogram in up to 10 clusters and make row annotation bar,\n",
    "        # also order matrix by dendogram clustering and annotate with cluster numbers\n",
    "        den_genes <- as.dendrogram(clust_genes)\n",
    "        den_split <- color_branches(den_genes, k=clust, col=wes_palette(\"Darjeeling1\", type=\"continuous\", n=10)) # colour branches\n",
    "\n",
    "        # get dendogram order & colours\n",
    "        col_labels <- get_leaves_branches_col(den_split)\n",
    "        col_labels <- col_labels[order(order.dendrogram(den_split))]\n",
    "\n",
    "        row_side_bar <- data.frame(col=col_labels)\n",
    "        row_side_bar <- data.frame(lapply(row_side_bar, as.character), stringsAsFactors=FALSE)\n",
    "\n",
    "        # get unique clusters & label them numerically\n",
    "        a <- as.data.frame(unique(row_side_bar))\n",
    "#         rownames(a) <- seq(length=nrow(a))\n",
    "        a$cluster <-  seq(length=nrow(a))\n",
    "\n",
    "        row_side_bar = plyr::join(row_side_bar, a, type=\"left\") # merge but keep order of clusters\n",
    "        row_side_bar$col <- NULL # drop extra col\n",
    "        \n",
    "    #         clust_matrix <- matrix[den_split$order,] # order matrix as in heatmap\n",
    "        clust_matrix <- matrix[clust_genes$labels, ]\n",
    "        clust_matrix <- cbind(clust_matrix, cluster=row_side_bar) # add RowSideColor clusters to matrix\n",
    "        clust_df <- data.frame(clust_matrix) # convert back to df\n",
    "        clust_df <- cbind(clust_df, gene_name=as.character(rownames(clust_matrix))) # get gene names\n",
    "\n",
    "        # create annotation bar of clusters\n",
    "        cluster_pal = c(wes_palette(\"Royal1\"), \"lightblue2\", \"#5BBCD6\", \"gold\", \"thistle\", \"lightgray\", \"darkseagreen\", \"skyblue\", \"indianred\")\n",
    "        cluster_pal = rep(cluster_pal, 5) # ensure palette is long enough\n",
    "        row_side_bar$cluster <- as.character(row_side_bar$cluster)\n",
    "        ha2 = HeatmapAnnotation(df=row_side_bar, \n",
    "                                col=list(cluster=c(\"1\"=cluster_pal[[1]], \"2\"= cluster_pal[[2]], \n",
    "                                                   \"3\"=cluster_pal[[3]], \"4\"=cluster_pal[[4]], \n",
    "                                                   \"5\"=cluster_pal[[5]], \"6\"=cluster_pal[[6]],\n",
    "                                                   \"7\"=cluster_pal[[7]], \"8\"=cluster_pal[[8]],\n",
    "                                                   \"9\"=cluster_pal[[9]], \"10\"=cluster_pal[[10]])), which=\"row\")\n",
    "    }\n",
    "    else {\n",
    "        den_genes <- as.dendrogram(clust_genes)\n",
    "        ha2 <- NULL\n",
    "        clust_matrix <- matrix\n",
    "    }\n",
    "\n",
    "    if(dist_method == \"minkowski\"){\n",
    "         p=1.44\n",
    "         dist_samples <- dist(t(matrix), method=dist_method, p)\n",
    "        }\n",
    "    else{\n",
    "         dist_samples <- dist(t(matrix), method=dist_method)\n",
    "        }\n",
    "\n",
    "    clust_samples <- hclust(dist_samples, method=clust_method)\n",
    "    den_samples <- as.dendrogram(clust_samples)\n",
    "\n",
    "    if (length(order) > 0){\n",
    "        # explicitly set optimal order of dendogram\n",
    "        den_samples <- den_samples %>% rotate(as.vector(order))\n",
    "    }\n",
    "    else{\n",
    "        den_samples <- den_samples %>% sort(type = \"labels\") # logically order labels\n",
    "    }\n",
    "\n",
    "    # label selection of genes\n",
    "    gene_labels <- rowAnnotation(link=row_anno_link(at=subset, labels=labels), width=unit(1, \"cm\") + max_text_width(labels))\n",
    "\n",
    "    # get range for colour palette\n",
    "    low <- quantile(matrix, low_opt)[[1]]\n",
    "    high <- quantile(matrix, high_opt)[[1]]\n",
    "    \n",
    "    if (med_opt==FALSE){\n",
    "    #             med <- mean(matrix)\n",
    "        med <- (high+low)/2\n",
    "    } else{\n",
    "            med <- quantile(matrix, med_opt)[[1]]\n",
    "    }\n",
    "    \n",
    "    # sample annotation bar\n",
    "    names <- unique(sampleinfo$condition)\n",
    "    cols <- wes_palette(\"Moonrise3\")[1:length(unique(sampleinfo$condition))]\n",
    "    named_vector1 <- setNames(cols, names)\n",
    "\n",
    "    names2 <- unique(sampleinfo$treatment) \n",
    "    cols2 <- wes_palette(\"FantasticFox1\")[1:length(unique(sampleinfo$treatment))]\n",
    "    named_vector2 <- setNames(cols2, names2)\n",
    "\n",
    "\n",
    "#     names3 <- unique(sampleinfo$treatment) \n",
    "#     cols3 <-  wes_palette(\"IsleofDogs1\")[1:length(unique(sampleinfo$group))]\n",
    "#     named_vector3 <- setNames(cols3, names3)\n",
    "\n",
    "#     ha = HeatmapAnnotation(group=sampleinfo$group, condition=sampleinfo$condition, treatment=sampleinfo$treatment, \n",
    "#                            col=list(group=named_vector1, condition=named_vector2, treatment=named_vector3))\n",
    "\n",
    "    ha = HeatmapAnnotation(condition=sampleinfo$condition, treatment=sampleinfo$treatment,\n",
    "                           col=list(condition=named_vector1, treatment=named_vector2))\n",
    "\n",
    "\n",
    "    \n",
    "    if (row_clust==FALSE){\n",
    "        den_genes <- FALSE\n",
    "    }\n",
    "\n",
    "    if (col_clust==FALSE){\n",
    "        den_samples <- FALSE\n",
    "    }\n",
    "\n",
    "    if (legend_label!=FALSE){\n",
    "        legend_name = legend_label\n",
    "    }\n",
    "    \n",
    "    ht = Heatmap(matrix, \n",
    "            col = colorRamp2(c(low, med, high), wes_palette(\"Zissou1\", type=\"continuous\", n=3), space=\"LAB\"),\n",
    "            name = legend_name,\n",
    "            cluster_rows = den_genes,\n",
    "            cluster_columns = den_samples,\n",
    "            column_dend_reorder = FALSE,\n",
    "            row_dend_reorder = FALSE,\n",
    "            column_dend_height = unit(1, \"cm\"),\n",
    "            row_dend_width = unit(2, \"cm\"),\n",
    "            show_row_names = show_rnames, \n",
    "            row_names_gp = gpar(fontsize=lsize),\n",
    "            show_column_names = FALSE, \n",
    "            heatmap_legend_param=list(legend_direction = \"horizontal\", \n",
    "                                      color_bar = \"continuous\",\n",
    "                                      legend_width = unit(5, \"cm\"), \n",
    "                                      title_position = \"lefttop\"),\n",
    "            show_column_dend = TRUE,\n",
    "            show_row_dend = TRUE,\n",
    "            column_title=title,\n",
    "            top_annotation=ha)\n",
    "\n",
    "    hmap <- draw(ht + ha2 + gene_labels, heatmap_legend_side = \"bottom\")\n",
    "    \n",
    "    return(list(hmap, clust_df, den_samples))\n",
    "}\n",
    "\n",
    "SigVstCounts <- function(DESEQ2_results, VSTcounts, gene_info, fdr=0.05, fc=0.58, index=\"gene_id\"){\n",
    "    # get VST normalised counts of significantly differentially expressed genes\n",
    "    # from DESEQ2 results object\n",
    "    \n",
    "    # get sig DE genes\n",
    "    if (fc != 0){\n",
    "        sigvar_genes <- rownames(DESEQ2_results)[DESEQ2_results$padj < fdr \n",
    "                                             & !is.na(DESEQ2_results$padj) \n",
    "                                             & DESEQ2_results$log2FoldChange > fc | \n",
    "                                             DESEQ2_results$padj < fdr \n",
    "                                             & !is.na(DESEQ2_results$padj) \n",
    "                                             & DESEQ2_results$log2FoldChange < -fc] \n",
    "    } else {\n",
    "        # don't filter on fold change, \n",
    "        # e.g. if using LRT test where FC isn't reflective of hypothesis tested\n",
    "        sigvar_genes <- rownames(DESEQ2_results)[DESEQ2_results$padj < fdr \n",
    "                                             & !is.na(DESEQ2_results$padj)]\n",
    "    }\n",
    "\n",
    "    sigVst <- VSTcounts[sigvar_genes,] # subset VST counts by DE genes\n",
    "    sigVst <- as.data.frame(sigVst)\n",
    "    \n",
    "    # subset counts by geneset (protein coding)\n",
    "    sigVst$gene_id <- rownames(sigVst)\n",
    "    sigVst <- merge(sigVst, gene_info[c(\"gene_id\", \"gene_name\")], by=\"gene_id\")\n",
    "\n",
    "    if (index == \"gene_id\"){\n",
    "        sigVst$gene_name <- NULL\n",
    "        rownames(sigVst) <- sigVst$gene_id\n",
    "        sigVst$gene_id <- NULL\n",
    "    }else if (index==\"gene_name\"){\n",
    "        sigVst$gene_id <- NULL\n",
    "        sigVst <- sigVst[!duplicated(sigVst$gene_name), ]\n",
    "        rownames(sigVst) <- sigVst$gene_name\n",
    "        sigVst$gene_name <- NULL\n",
    "    }else{\n",
    "        print(\"specify either gene_id or gene_name for rownames\")\n",
    "    }\n",
    "    \n",
    "    \n",
    "    print(paste(\"No. DE genes:\", as.character(nrow(sigVst))))\n",
    "\n",
    "    return(sigVst)   \n",
    "}\n",
    "\n",
    "go_analysis <- function(goi, bg, ontologies=c(\"BP\", \"MF\", \"CC\")) {\n",
    "    # iterate thorugh list of clusters, getting enriched BP, MF, & BP GO terms for each\n",
    "    # store data in df, annotated by ontology and cluster\n",
    "\n",
    "    clusters <- unique(goi$cluster)\n",
    "    \n",
    "    x <- 0\n",
    "    for (o in ontologies) {\n",
    "        x <- x + 1\n",
    "        y <- 0\n",
    "        for (c in clusters) {\n",
    "            y <- y + 1\n",
    "\n",
    "            clust <- goi[goi$cluster == c, ]$gene_id # get genes in cluster\n",
    "            gene_list <- factor(as.integer(bg %in% clust)) # annotate gene universe with goi's from cluster\n",
    "            names(gene_list) <- bg # add names to vector\n",
    "\n",
    "            # create topGO data object\n",
    "            GO <- new(\"topGOdata\", ontology = o, allGenes = gene_list, \n",
    "                          annotationFun = annFUN.org, o, mapping=\"org.Mm.eg.db\", ID=\"ensembl\")\n",
    "\n",
    "            # Fisher's exact test for enrichment of genes in cluster vs background\n",
    "            GOres <- runTest(GO, algorithm=\"classic\", statistic=\"fisher\")\n",
    "\n",
    "            # algorithm = \"classic\" - GO hierarchy isn't taken into account\n",
    "            # try conditional enrichment to control for redundancy (enrichment of parent terms)\n",
    "            GOres_weighted <- runTest(GO, algorithm=\"weight01\", statistic=\"fisher\")\n",
    "            \n",
    "            # extract results for each cluster\n",
    "            if (y == 1) {\n",
    "                res <- GenTable(GO, weightFisher=GOres_weighted, orderBy=\"classicFisher\", classicFisher=GOres)\n",
    "                res$cluster <- c\n",
    "                res$ontology <- o\n",
    "                res$padj_weightFisher <- p.adjust(res$weightFisher, method=\"BH\")\n",
    "                res$padj <- p.adjust(res$classicFisher, method=\"BH\")\n",
    "\n",
    "                res <- res[order(res$padj), ]\n",
    "                \n",
    "            } else {\n",
    "                df <- GenTable(GO, weightFisher=GOres_weighted, orderBy=\"classicFisher\", classicFisher=GOres)\n",
    "                df$cluster <- c\n",
    "                df$ontology <- o\n",
    "                df$padj_weightFisher <- p.adjust(df$weightFisher, method=\"BH\")\n",
    "                df$padj <- p.adjust(df$classicFisher, method=\"BH\")\n",
    "\n",
    "                df <- df[order(df$padj), ]\n",
    "\n",
    "                res <- rbind(res, df)\n",
    "            }\n",
    "\n",
    "        }\n",
    "        \n",
    "    if (x == 1){\n",
    "        merge <- res\n",
    "    } else {\n",
    "        merge <- rbind(merge, res)\n",
    "    }\n",
    "    \n",
    "    }\n",
    "    result <- merge[order(merge$cluster), ]\n",
    "    \n",
    "    return(result)\n",
    "}\n",
    "\n",
    "# Set ggplot theme\n",
    "theme_set(theme_notebook(base_size=18))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info = fetch_DataFrame('''select * from sample_info''', db)\n",
    "sample_info.index = sample_info[\"sample_id\"]\n",
    "sample_info.index.name = None\n",
    "sample_info.head(len(sample_info))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i sample_info -o Palette\n",
    "\n",
    "# get pretty colours\n",
    "Palette <- wes_palette(sample(names(wes_palettes), 1))\n",
    "\n",
    "if (length(Palette) < length(unique(sample_info$category))){\n",
    "    while (length(Palette) < length(unique(sample_info$category))){\n",
    "        pal <- wes_palette(sample(names(wes_palettes), 1))\n",
    "        Palette <- unique(c(Palette, pal))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get counts\n",
    "counts = fetch_DataFrame('''select * from featureCounts''', db)\n",
    "cols = [x for x in sample_info[\"sample_id\"]] + [\"Geneid\"]\n",
    "counts = counts[cols]\n",
    "counts.rename(columns={\"Geneid\":\"gene_id\"}, inplace=True)\n",
    "counts.set_index(counts[\"gene_id\"], inplace=True) # set index\n",
    "counts.drop(\"gene_id\", axis=1, inplace=True)\n",
    "counts.index.name = None\n",
    "\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene_ids & gene_names for protein coding genes\n",
    "\n",
    "statement = '''select distinct b.contig, b.start, b.end, a.gene_id, \n",
    "            a.gene_name from gene_info a, geneset_all_gtf_genome_coordinates \n",
    "            b where a.gene_id = b.gene_id and transcript_biotype = \"protein_coding\"''' \n",
    "\n",
    "gene_info = fetch_DataFrame(statement, anndb)\n",
    "\n",
    "# gene_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<br>\n",
    "\n",
    "## DESeq2 Analysis\n",
    "- DESeq2 analysis design is indicated by design column in sample information table\n",
    "- Variance stabilised (VST) counts of all differentially expressed (LRT test, padj < 0.05) genes are used for PCA, tSNE, and heatmap\n",
    "\n",
    "#### Plots:\n",
    "- standard deviation for VST transformed data vs mean (across samples)\n",
    "- dispersion estimate plot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " %%R -i counts,sample_info -w 600\n",
    "\n",
    "# define a data frame with information about the samples (i.e the columns of the countData)\n",
    "rownames(sample_info) <- sample_info$sample_id\n",
    "sample_info$sample_id <- NULL\n",
    "\n",
    "# reorder the columnData to match the count matrix:\n",
    "sample_info <- sample_info[colnames(counts),]\n",
    "\n",
    "# build a DESeqDataSet from the count matrix\n",
    "dds <- DESeqDataSetFromMatrix(countData=counts, \n",
    "                              colData=sample_info, \n",
    "                              design= ~ category )\n",
    "\n",
    "dds <- estimateSizeFactors(dds)\n",
    "\n",
    "# apply VST\n",
    "vstMat <- assay(varianceStabilizingTransformation(dds, blind=TRUE, fitType=\"local\")) \n",
    "colnames(vstMat) <- rownames(sample_info)\n",
    "notAllZero <- (rowMeans(vstMat)> min(vstMat))\n",
    "\n",
    "meanSdPlot(vstMat[notAllZero,], ylab=\"sd, vst\",plot=T) \n",
    "\n",
    "# run DESeq2\n",
    "dds <- DESeq(dds,  fitType=\"local\") # have check both and local provides best fit\n",
    "dds_lrt <- DESeq(dds,  fitType=\"local\", test=\"LRT\", reduced=~1)\n",
    "\n",
    "res_lrt <- results(dds_lrt)\n",
    "res <- results(dds)\n",
    "\n",
    "plotDispEsts(dds) # check dispersion estimate fit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cooks Distance Cutoff\n",
    "- DESeq function calculates Cooks distance (diagnostic test for outliers) for every gene and sample\n",
    "- Cooks distance is a measure of how much a single sample is influencing fitted coefficients for a gene, large values indicate an outlier count\n",
    "- results function automatically flags genes with high Cooks values for samples w/ >3 replicates. P & padj values for these genes are set to NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 1200 -h 400\n",
    "\n",
    "# DESeq function calculates Cooks distance (diagnostic test for outliers) for every gene and sample. \n",
    "# Cooks distance is a measure of how much a single sample is influencing fitted coefficients for a gene, large values indicate an outlier count\n",
    "# results function automatically flags genes with high Cooks values for samples w/ >3 replicates. P & padj values for these genes are set to NA\n",
    "# Filtering on genes by Cooks distance can be turned off with results(dds, cooksCutoff=FALSE)\n",
    "\n",
    "cooks <- melt(as.data.frame(log10(assays(dds)[[\"cooks\"]])))\n",
    "cooks <- merge(cooks, sample_info, by.x=\"variable\", by.y=0, all=T)\n",
    "\n",
    "p1 <- ggplot(cooks, aes(y=value, x=variable, fill=category)) +  \n",
    "        geom_boxplot(colour=\"black\")  + \n",
    "        labs(y=\"Cooks Distance (log10)\", x=\"\") +\n",
    "        scale_fill_manual(values=Palette) +\n",
    "        theme_notebook() +\n",
    "#         theme(axis.text.x=element_text(angle=45, hjust=1)) +\n",
    "        theme(axis.text.x=element_blank()) +\n",
    "        geom_hline(lty=\"dashed\", col=\"black\", yintercept=mean(cooks$value, na.rm=T))\n",
    "\n",
    "p1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "### Count normalisation \n",
    "\n",
    "\n",
    "Effect of normalisation on counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 1200 -h 600 -o filt_norm_counts\n",
    "\n",
    "# raw counts\n",
    "raw <- melt(t(counts(dds)))\n",
    "raw <- merge(raw, sample_info, by.x=\"Var1\", by.y=0, all=T)\n",
    "p1 <- ggplot(raw, aes(y=log2(value), x=Var1, fill=category)) + \n",
    "        geom_boxplot(colour=\"black\") + \n",
    "        theme_notebook() +\n",
    "        theme(axis.text.x=element_blank()) +\n",
    "        labs(title=\"Raw counts\", y=\"log2(counts)\", x=\"\") + \n",
    "        scale_fill_manual(values=Palette) \n",
    "\n",
    "\n",
    "\n",
    "# norm counts = counts(dds) / sizeFactors(dds)\n",
    "norm <- melt(t(counts(dds,normalized=T)))\n",
    "norm <- merge(norm, sample_info, by.x=\"Var1\", by.y=0, all=T)\n",
    "p2 <- ggplot(norm, aes(y=log2(value), x=Var1, fill=category)) + \n",
    "        geom_boxplot(colour=\"black\") + \n",
    "        theme_notebook() +\n",
    "        theme(axis.text.x=element_blank()) +\n",
    "        labs(title=\"Normalised counts\", y=\"\", x=\"\") + \n",
    "        scale_fill_manual(values=Palette)\n",
    "\n",
    "\n",
    "\n",
    "# VST counts\n",
    "colnames(vstMat) <- dds@colData@rownames\n",
    "data <-melt(vstMat[notAllZero,])\n",
    "data <- merge(data, sample_info, by.x=\"Var2\", by.y=0, all=T)\n",
    "\n",
    "p3 <- ggplot(data, aes(y=value, x=Var2, fill=category)) + \n",
    "        geom_boxplot(colour=\"black\")  + \n",
    "        theme_notebook() +\n",
    "        theme(axis.text.x=element_blank()) +\n",
    "        labs(title=\"VST counts\", y=\"\", x=\"\") + \n",
    "        scale_fill_manual(values=Palette) \n",
    "\n",
    "# normalised counts with excluded counts (padj == NA) removed\n",
    "# counts excluded by: Cooks cutoff & independent filtering\n",
    "filtered_genes <- subset(res, res@listData$padj != \"NA\")@\"rownames\" # get sig DE gene names\n",
    "filt_norm_counts <- subset(melt(t(counts(dds,normalized=T))), Var2 %in% filtered_genes)\n",
    "filt_norm_counts_info <- merge(filt_norm_counts, sample_info, by.x=\"Var1\", by.y=0, all=T)\n",
    "\n",
    "p4 <- ggplot(filt_norm_counts_info, aes(y=log2(value), x=Var1, fill=category)) + \n",
    "        geom_boxplot(colour=\"black\") + \n",
    "        theme_notebook() +\n",
    "        theme(axis.text.x=element_blank()) +\n",
    "        labs(title=\"Filt. normalised counts\", y=\"log2(counts)\", x=\"\") + \n",
    "        scale_fill_manual(values=Palette) \n",
    "\n",
    "get_legend <- function(a.gplot){ \n",
    "  tmp <- ggplot_gtable(ggplot_build(a.gplot)) \n",
    "  leg <- which(sapply(tmp$grobs, function(x) x$name) == \"guide-box\") \n",
    "  legend <- tmp$grobs[[leg]] \n",
    "  return(legend)}\n",
    "                      \n",
    "key <- get_legend(p4 + theme(legend.direction=\"horizontal\"))\n",
    "\n",
    "grid.arrange(p1 + theme(legend.position=\"none\"), \n",
    "             p2 + theme(legend.position=\"none\"), \n",
    "             p4 + theme(legend.position=\"none\"), \n",
    "             p3 + theme(legend.position=\"none\"), bottom=key, ncol=2, nrow=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rearrange data -> by gene\n",
    "filt_norm_counts = filt_norm_counts.pivot(\"Var2\", \"Var1\", \"value\")\n",
    "filt_norm_counts.index.name = None\n",
    "# filt_norm_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relative Log Expression (RLE)\n",
    "- RLE is implemented in edgeR and DESeq packages - normalises samples for size factors\n",
    "- Briefly, the gene expression in each sample is normalised to the median expression of that gene across all samples & counts are log2 transformed\n",
    "- RLE should be centered over 0, dispersion from this indicates poor quality normalisation\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 1200 -h 600 -i filt_norm_counts\n",
    "\n",
    "# relative log2 expression\n",
    "# Relative Log Expression” (RLE) is implemented in edgeR and DESeq packages - normalises samples for size factors\n",
    "# good description of procedure here: http://www.nathalievilla.org/doc/pdf/tutorial-rnaseq.pdf\n",
    "# Briefly, the gene expression in each sample is normalised to the median expression of that gene across all samples & counts are log2 transformed\n",
    "# RLE should be centered over 0, dispersion from this indicates poor quality normalisation\n",
    "\n",
    "# raw counts\n",
    "rawc <- apply(counts(dds),1,median)\n",
    "rle_raw <- log2(counts(dds)/rawc)\n",
    "rlem_raw <- melt(t(rle_raw))\n",
    "rlem_raw <- merge(rlem_raw, sample_info, by.x=\"Var1\", by.y=0, all=T)\n",
    "# head(rlem_raw)\n",
    "\n",
    "rle_raw_bp <- ggplot(rlem_raw, aes(y=value, x=Var1, fill=category)) + geom_boxplot(colour=\"black\") + \n",
    "                theme_notebook() +\n",
    "                theme(axis.text.x=element_blank()) + \n",
    "                geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                labs(title=\"Raw counts\", y=\"Relative Log Expression\", x=\"\") +\n",
    "                scale_fill_manual(values=c(Palette)) \n",
    "\n",
    "\n",
    "# norm counts\n",
    "normc <- apply(counts(dds, normalized=TRUE),1,median)\n",
    "rle_norm <- log2(counts(dds, normalized=TRUE)/normc)\n",
    "rlem_norm <- melt(t(rle_norm))\n",
    "rlem_norm <- merge(rlem_norm, sample_info, by.x=\"Var1\", by.y=0, all=T)\n",
    "# head(rlem_norm)\n",
    "\n",
    "rle_norm_bp <- ggplot(rlem_norm, aes(y=value, x=Var1, fill=category)) + geom_boxplot(colour=\"black\") + \n",
    "                theme_notebook() +\n",
    "                theme(axis.text.x=element_blank()) + \n",
    "                geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                labs(title=\"Normalised counts\", y=\"\", x=\"\") +\n",
    "                scale_fill_manual(values=c(Palette))\n",
    "\n",
    "# Norm counts with filtering (Cooks distance, independent filtering)\n",
    "filtc <- apply(filt_norm_counts,1,median)\n",
    "rle_filt <- log2(filt_norm_counts/filtc)\n",
    "rlem_filt <- melt(t(rle_filt))\n",
    "rlem_filt <- merge(rlem_filt, sample_info, by.x=\"Var1\", by.y=0, all=T)\n",
    "# head(rlem_filt)\n",
    "\n",
    "rle_filt_bp <- ggplot(rlem_filt, aes(y=value, x=Var1, fill=category)) + geom_boxplot(colour=\"black\") + \n",
    "                theme_notebook() +\n",
    "                theme(axis.text.x=element_blank()) + \n",
    "                geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                labs(title=\"Filt. normalised counts\", y=\"Relative Log Expression\", x=\"\") +\n",
    "                scale_fill_manual(values=c(Palette)) \n",
    "\n",
    "# VST counts\n",
    "vstc <- apply(vstMat,1,median)\n",
    "rle_vst <- log2(vstMat/vstc)\n",
    "rlem_vst <- melt(t(rle_vst))\n",
    "rlem_vst <- merge(rlem_vst, sample_info, by.x=\"Var1\", by.y=0, all=T)\n",
    "\n",
    "rle_vst_bp <- ggplot(rlem_vst, aes(y=value, x=Var1, colour=category)) + geom_boxplot() + \n",
    "                theme_notebook() +\n",
    "                theme(axis.text.x=element_blank()) + \n",
    "                geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                labs(title=\"VST counts\", y=\"\", x=\"\") +\n",
    "                scale_colour_manual(values=c(Palette)) \n",
    "\n",
    "key <- get_legend(rle_raw_bp + theme(legend.direction=\"horizontal\"))\n",
    "\n",
    "grid.arrange(rle_raw_bp + theme(legend.position=\"none\"), \n",
    "             rle_norm_bp + theme(legend.position=\"none\"), \n",
    "             rle_filt_bp + theme(legend.position=\"none\"), \n",
    "             rle_vst_bp + theme(legend.position=\"none\"), ncol=2, nrow=2, bottom=key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o sigVst,vstDF -i sample_info,gene_info\n",
    "\n",
    "sigVst <- SigVstCounts(res_lrt, vstMat, gene_info, fdr=0.05, fc=0)\n",
    "\n",
    "# also export all counts to df\n",
    "vstDF <- as.data.frame(vstMat)\n",
    "vstDF$gene_id <- rownames(vstMat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 1200 -h 450\n",
    "                      \n",
    "pca <- prcomp(t(sigVst), scale=FALSE)\n",
    "\n",
    "pca_plots <- ggplot_prcomp(pca, \n",
    "             plots=list(\"A\"=c(\"PC1\",\"PC2\"), \"B\"=c(\"PC3\", \"PC4\")),\n",
    "             sample_information=sample_info, \n",
    "             color=\"condition\",\n",
    "             label=\"replicate\",\n",
    "             shape=\"treatment\",\n",
    "             size=6,\n",
    "             nudge_scale_factor=30) \n",
    "\n",
    "a <- pca_plots$A\n",
    "b <- pca_plots$B\n",
    "c <- pca_plots$scree\n",
    "\n",
    "a = a + scale_colour_manual(values=Palette) +\n",
    "        guides(colour=guide_legend(override.aes=list(size=8)), shape=guide_legend(override.aes=list(size=8)))\n",
    "               \n",
    "b = b + scale_colour_manual(values=Palette) \n",
    "\n",
    "c = c \n",
    "\n",
    "get_legend <- function(a.gplot){ \n",
    "tmp <- ggplot_gtable(ggplot_build(a.gplot)) \n",
    "leg <- which(sapply(tmp$grobs, function(x) x$name) == \"guide-box\") \n",
    "legend <- tmp$grobs[[leg]] \n",
    "return(legend)}\n",
    "\n",
    "a <- a + theme(legend.direction=\"horizontal\", legend.box=\"horizontal\")\n",
    "legend <- get_legend(a)\n",
    "                      \n",
    "grid.arrange(top=textGrob(\"PCA of VST Counts at Differentially Accessible Peaks\", \n",
    "                          gp=gpar(fontface=\"bold\", fontfamily=\"Helvetica\", fontsize=23)),\n",
    "             a + theme(legend.position=\"none\"), b + theme(legend.position=\"none\"), c, \n",
    "             bottom=legend, ncol=3, nrow=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 450 -h 400\n",
    "\n",
    "library(Rtsne)\n",
    "tsne_out = Rtsne(t(sigVst), pca=T, perplexity=3)\n",
    "\n",
    "tsne_df <- as.data.frame(tsne_out$Y)\n",
    "rownames(tsne_df) <- colnames(t(sample_info))\n",
    "colnames(tsne_df) <- c(\"tSNE1\", \"tSNE2\")\n",
    "\n",
    "tsne_df <- merge(tsne_df, sample_info, by=0)\n",
    "\n",
    "tsne_plot <- ggplot(tsne_df, aes(y=tSNE1, x=tSNE2, colour=category, shape=factor(replicate))) + \n",
    "                    geom_point(size=6) + \n",
    "                    theme_notebook() +\n",
    "                    scale_colour_manual(values=Palette) +\n",
    "                    scale_shape_manual(values=c(16,17,22,23,25),name=\"replicate\")\n",
    "\n",
    "grid.arrange(top=textGrob(\"tSNE of VST Counts at \\nDifferentially Accessible Peaks\", \n",
    "             gp=gpar(fontface=\"bold\", fontfamily=\"Helvetica\", fontsize=18)), \n",
    "             tsne_plot, ncol=1, nrow=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def upload2csvdb(df, tablename, db):\n",
    "    '''upload df to sqlite db'''\n",
    "    \n",
    "    connect = sqlite3.connect(db)\n",
    "    \n",
    "    df.to_sql(tablename, connect, if_exists=\"replace\", index=False)\n",
    "    \n",
    "\n",
    "upload2csvdb(sigVst, \"DESeq2_LRT_all_samples\", \"./csvdb\")\n",
    "upload2csvdb(vstDF, \"VST_counts\", \"./csvdb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "### Pearson correlation\n",
    "- based on Manhattan distances between samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 500 -h 500\n",
    "\n",
    "cm <- data.matrix(sigVst)\n",
    "m <- cor(cm, method=\"pearson\", use=\"all\")\n",
    "\n",
    "distr <- dist(m, method=\"manhattan\")\n",
    "clustr <- hclust(distr, method=\"ward.D2\")\n",
    "dendr <- as.dendrogram(clustr)\n",
    "dendr <- dendr %>% sort(type=\"labels\")\n",
    "\n",
    "distc <- dist(t(m), method=\"manhattan\")\n",
    "clustc <- hclust(distc, method=\"ward.D2\")\n",
    "dendc <- as.dendrogram(clustc)\n",
    "dendc <- rev(dendc) %>% sort(type=\"labels\")\n",
    "\n",
    "p2 <- Heatmap(m,\n",
    "#        split=4,\n",
    "       col = colorRamp2(c(min(m), median(m), max(m)), wes_palette(\"Zissou1\", type=\"continuous\", n=3)),\n",
    "       cluster_rows=dendr,\n",
    "       cluster_columns=dendc,\n",
    "       column_dend_reorder = FALSE,\n",
    "       row_dend_reorder = FALSE,\n",
    "       column_dend_height = unit(2, \"cm\"),\n",
    "       row_dend_width = unit(2, \"cm\"),\n",
    "       row_names_gp=gpar(fontsize=12),\n",
    "       column_names_gp=gpar(fontsize=12),\n",
    "       name=\"Pearson Correlation:\",\n",
    "       heatmap_legend_param=list(legend_direction=\"horizontal\", \n",
    "#                                   at=c(0.9, 1), \n",
    "                                  color_bar = \"continuous\",\n",
    "                                  legend_width = unit(5, \"cm\"), \n",
    "                                  title_position = \"lefttop\",\n",
    "                                  title_gp=gpar(fontsize=14),\n",
    "                                  labels_gp=gpar(fontsize=14)),\n",
    "       )\n",
    "\n",
    "draw(p2, heatmap_legend_side = \"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "### Hierarchical clustering \n",
    "- based on expression of DE genes\n",
    "- VST counts and row scaled (i.e. min to max expression for each gene) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -h 600 -w 500\n",
    "\n",
    "# get significat LRT genes in format for heatmap\n",
    "sigVst <- SigVstCounts(res_lrt, vstMat, gene_info, fdr=0.05, fc=0, index=\"gene_name\")\n",
    "sigVst$gene_name <- rownames(sigVst)\n",
    "\n",
    "# head(sigVst)\n",
    "hmap <- plot_heatmap(sigVst, sample_info, Palette, scale_data=\"none\", label=\"top_var\", clust=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -h 600 -w 500\n",
    "\n",
    "hmap <- plot_heatmap(sigVst, sample_info, Palette, scale_data=\"min_max\", label=\"top_var\", clust=6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene ontology for clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# get gene background, all expressed genes\n",
    "bg = counts.copy(deep=True)\n",
    "\n",
    "# get gene names\n",
    "bg[\"gene_id\"] = counts.index.values\n",
    "bg = pd.merge(bg, gene_info[[\"gene_id\", \"gene_name\"]], how=\"inner\", on=\"gene_id\")\n",
    "\n",
    "bg.set_index(\"gene_id\", inplace=True, drop=True)\n",
    "bg.drop(\"gene_name\", axis=1, inplace=True)\n",
    "bg.index.name = None #(inplace=True)\n",
    "\n",
    "# subset to detected genes\n",
    "# print(len(bg))\n",
    "bg = bg[(bg != 0).any(axis=1)]\n",
    "\n",
    "# gene ids only\n",
    "bg = bg.index.values\n",
    "\n",
    "# len(bg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# get genes of interest\n",
    "goi <- hmap[[2]]\n",
    "goi <- merge(goi[, c(\"gene_name\", \"cluster\")], gene_info[, c(\"gene_id\", \"gene_name\")], how=\"inner\", by=\"gene_name\")\n",
    "\n",
    "# head(goi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i bg\n",
    "\n",
    "go <- go_analysis(goi, bg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "saveRDS(go, file=\"LRT_heatmap_clusters_GO.rds\")\n",
    "# go <- readRDS(\"LRT_heatmap_clusters_GO.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -h 800 -w 1200\n",
    "\n",
    "library(stringr)\n",
    "plot_go <- function(df, pval=0.1, top=10, title=\"\"){\n",
    "    df$Fold <- df$Significant / df$Expected\n",
    "\n",
    "    df <- subset(df, padj < pval)\n",
    "        \n",
    "    n <- 0\n",
    "    for (c in unique(df$cluster)){\n",
    "        n <- n +1\n",
    "        cat <- subset(df, cluster == c)\n",
    "        cat <- cat[order(cat$padj, decreasing=FALSE),  ]\n",
    "        cat <- head(cat, top)\n",
    "        \n",
    "        if (n==1){\n",
    "            res <- cat\n",
    "        } else {\n",
    "            res <- rbind(res, cat)\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    res <- res[order(res$cluster), ]\n",
    "\n",
    "    clust_pal <- c(wes_palette(\"Royal1\"), \"lightblue2\", \"#5BBCD6\", \"gold\", \"thistle\", \"lightgray\", \"darkseagreen\", \"skyblue\", \"indianred\")\n",
    "\n",
    "    p1 <- ggplot(res, aes(y=Fold, x=Term, fill=as.factor(cluster), alpha=-log10(padj))) + \n",
    "            geom_bar(stat=\"identity\", position=\"dodge\", colour=\"black\") + \n",
    "            geom_text(aes(label=Significant, y=Fold+0.2,), colour=\"black\", position=position_dodge(width=1), alpha=1, size=4) +\n",
    "#             geom_text(aes(label=GO.ID, y=Fold/2,), colour=\"black\", position=position_dodge(width=1), alpha=1, size=4) +\n",
    "            scale_x_discrete(limits=unique(res$Term), labels=str_wrap(unique(res$Term), 50)) +\n",
    "            scale_fill_manual(values=clust_pal, name=\"Cluster\") +\n",
    "            scale_alpha(range=c(0.4, 1), name=\"-log10(FDR)\") +\n",
    "            coord_flip() +\n",
    "            labs(x=\"\", title=paste(\"Ontology:\", title)) +\n",
    "            theme_notebook() +\n",
    "            theme(legend.box=\"vertical\") +\n",
    "            geom_hline(yintercept=1, lty=\"dashed\") \n",
    "    \n",
    "    \n",
    "    return(p1)\n",
    "}\n",
    "\n",
    "for (o in unique(go$ontology)){\n",
    "    p1 <- plot_go(subset(go, ontology==o), top=5, title=o)\n",
    "    grid.arrange(p1, ncol=1, nrow=1)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DE Peaks - Contrasts\n",
    "* Pairwise comparisons of differentially accessible peaks\n",
    "* Adjusted p-value < 0.05 & fold change > 2\n",
    "* DESeq2 contrasts:\n",
    "    * Fold changes reported are relative to \"a\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o condition_pairs\n",
    "\n",
    "# get pairs of sample comparisons\n",
    "a <- unique(sample_info$category)\n",
    "b <- combn(a, 2, simplify=FALSE)\n",
    "condition_pairs <- as.data.frame(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if comparisons specified in pipeline.yml use these\n",
    "comps = opts[\"report\"][\"comparisons\"]\n",
    "if comps:\n",
    "    condition_pairs = pd.DataFrame.from_dict(comps, orient=\"index\", columns=[\"a\", \"b\"])\n",
    "    condition_pairs[\"comparison\"] = condition_pairs.index.values\n",
    "    condition_pairs.reset_index(inplace=True, drop=True)\n",
    "\n",
    "else: # use randomly generated comparisons\n",
    "    condition_pairs = condition_pairs.transpose()\n",
    "\n",
    "    condition_pairs.reset_index(drop=True, inplace=True)\n",
    "    condition_pairs.columns = [\"a\", \"b\"]\n",
    "    condition_pairs[\"comparison\"] = condition_pairs.apply(lambda x: '_'.join([str(x.a), \"VS\", str(x.b)]), axis=1)\n",
    "\n",
    "condition_pairs.head(len(condition_pairs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -h 550 -w 1200 -i condition_pairs\n",
    "\n",
    "MA <- function(df, title=\"\", gene_info=gene_info){\n",
    "    \n",
    "    df$gene_id <- rownames(df)\n",
    "    \n",
    "    df <- merge(df, gene_info[c(\"gene_id\", \"gene_name\")], by.x=0, by.y=\"gene_id\")\n",
    "    \n",
    "    MA <- ggplot(subset(df, padj > 0.05 | is.na(padj)), aes(y=log2FoldChange, x=log2(baseMean))) + \n",
    "            geom_point(alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange < 1 | padj < 0.05 & log2FoldChange > -1), \n",
    "                       aes(y=log2FoldChange, x=log2(baseMean)), alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange > 1 | padj < 0.05 & log2FoldChange < -1), \n",
    "                       aes(y=log2FoldChange, x=log2(baseMean)), alpha=0.4, colour=\"red3\") +\n",
    "            scale_y_continuous(limits=c(min(df$log2FoldChange)-2, max(df$log2FoldChange)+2)) +\n",
    "            scale_x_continuous(limits=c(0, max(log2(df$baseMean)))) +\n",
    "            theme_notebook() +\n",
    "            labs(y=\"log2(FoldChange)\", title=title) +\n",
    "            geom_hline(yintercept=0, lty=\"dashed\", col=\"black\") +\n",
    "            geom_label_repel(data=subset(df[df$padj < quantile(df$padj, 0.0005, na.rm=TRUE), ], padj < 0.05),\n",
    "                             aes(y=log2FoldChange, x=log2(baseMean), \n",
    "                             label=gene_name), size=5, \n",
    "                             label.padding=unit(0.3, \"lines\"), label.r=unit(0.5, \"lines\"), segment.color = \"gray20\", \n",
    "                             fill=\"gray20\", colour=\"white\", fontface=\"bold\", label.size=NA, alpha=0.8)\n",
    "\n",
    "    return(MA)\n",
    "}\n",
    "\n",
    "volcano <- function(df, title=\"\", gene_info=gene_info){\n",
    "      \n",
    "    df$gene_id <- rownames(df)\n",
    "    \n",
    "    df <- merge(df, gene_info[c(\"gene_id\", \"gene_name\")], by.x=0, by.y=\"gene_id\")\n",
    "    \n",
    "    volcano <- ggplot(subset(df, padj > 0.05 | is.na(padj)), aes(y=log2(padj), x=log2FoldChange)) + \n",
    "            geom_point(alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange < 1 | padj < 0.05 & log2FoldChange > -1), \n",
    "                       aes(y=log2(padj), x=log2FoldChange), alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange > 1 | padj < 0.05 & log2FoldChange < -1), \n",
    "                       aes(y=log2(padj), x=log2FoldChange), alpha=0.4, colour=\"red3\") +\n",
    "            labs(y=\"log2(FDR)\", title=title) +\n",
    "            scale_y_reverse(limits=c(0, min(log2(df$padj)))) +\n",
    "            scale_x_continuous(limits=c(min(df$log2FoldChange), max(df$log2FoldChange))) +\n",
    "            theme_notebook() +\n",
    "            geom_vline(xintercept=0, lty=\"dashed\", col=\"black\") +\n",
    "            geom_label_repel(data=subset(df[df$padj < quantile(df$padj, 0.0005, na.rm=TRUE), ], padj < 0.05),\n",
    "                             aes(y=log2(padj), x=log2FoldChange, \n",
    "                             label=gene_name), size=5, \n",
    "                             label.padding=unit(0.3, \"lines\"), label.r=unit(0.5, \"lines\"), segment.color = \"gray20\", \n",
    "                             fill=\"gray20\", colour=\"white\", fontface=\"bold\", label.size=NA, alpha=0.8)\n",
    "\n",
    "    return(volcano)\n",
    "}\n",
    "\n",
    "n = 0\n",
    "for (line in 1:nrow(condition_pairs)){\n",
    "    n = n + 1\n",
    "\n",
    "    sample = condition_pairs[n, ]\n",
    "    a <- as.character(sample$a)\n",
    "    b <- as.character(sample$b)\n",
    "    title <- as.character(sample$comparison)\n",
    "\n",
    "    plot <- as.data.frame(results(dds, contrast=c(\"category\", a, b)))\n",
    "    p1 <- (MA(plot, title, gene_info=gene_info))\n",
    "    p2 <- (volcano(plot, title, gene_info=gene_info))\n",
    "\n",
    "    grid.arrange(p1, p2, ncol=2, nrow=1)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary Table \n",
    "* Peaks with adjusted p-value < 0.05 and fold change > 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o sig_genes\n",
    "\n",
    "# get no. sig diff. accessible peaks\n",
    "\n",
    "n = 0\n",
    "contrasts = list()\n",
    "for (line in 1:nrow(condition_pairs)){\n",
    "    n = n + 1\n",
    "\n",
    "    sample = condition_pairs[n, ]\n",
    "\n",
    "    a <- as.character(sample$a)\n",
    "    b <- as.character(sample$b)\n",
    "    title <- as.character(sample$comparison)\n",
    "    \n",
    "    if (length(title)[1] > 0){\n",
    "        contrasts[[title]] <- c(a, b)\n",
    "    }\n",
    "\n",
    "    # get significant results\n",
    "    df <- as.data.frame(results(dds, contrast=c(\"category\", a, b)))\n",
    "    df$gene_id <- rownames(df)\n",
    "    df <- merge(df, gene_info[, c(\"gene_id\", \"gene_name\")], how=\"inner\", by=\"gene_id\")\n",
    "    res <- subset(df, padj < 0.05 & log2FoldChange > 1 |  padj < 0.05 & log2FoldChange < -1)\n",
    "    upreg <- subset(res, log2FoldChange > 0)\n",
    "    downreg <- subset(res, log2FoldChange < 0)\n",
    "    \n",
    "    if (n == 1) {\n",
    "        sig_genes <- as.data.frame(c(length(res$padj), length(upreg$padj), length(downreg$padj), title), byrow=F)\n",
    "        sig_genes <- t(sig_genes)\n",
    "    } else{\n",
    "        df <- as.data.frame(c(length(res$padj), length(upreg$padj), length(downreg$padj), title), byrow=F)\n",
    "        df <- t(df)\n",
    "        sig_genes <- rbind(sig_genes, df)\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "sig_genes <- as.data.frame(sig_genes)\n",
    "rownames(sig_genes) <- NULL\n",
    "colnames(sig_genes) <- c(\"No. genes\", \"Upreg\", \"Downreg\", \"Contrast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_genes[\"No. genes\"] = sig_genes[\"No. genes\"].apply(lambda x: int(x))\n",
    "sig_genes.sort_values(\"No. genes\", ascending=False).head(len(sig_genes))\n",
    "\n",
    "sig_genes.head(len(sig_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "### DESeq2 results are available at:\n",
    "* Differentially accessible peaks saved in \"DESeq2_results\" table in csvdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o de_genes,all_genes\n",
    "\n",
    "# make table of all significant results for each comparison & upload to db\n",
    "\n",
    "n <- 0\n",
    "contrasts <- list()\n",
    "for (line in 1:nrow(condition_pairs)){\n",
    "    n <- n + 1\n",
    "\n",
    "    sample <- condition_pairs[n, ]\n",
    "\n",
    "    a <- as.character(sample$a)\n",
    "    b <- as.character(sample$b)\n",
    "    title <- as.character(sample$comparison)\n",
    "    \n",
    "    if (length(title)[1] > 0){\n",
    "        contrasts[[title]] <- c(a, b)\n",
    "    }\n",
    "\n",
    "    # get significant results\n",
    "    res <- results(dds, contrast=c(\"category\", a, b))\n",
    "    sig_res <- subset(res, padj < 0.05 & log2FoldChange > 1 |  padj < 0.05 & log2FoldChange < -1)\n",
    "    upreg <- as.data.frame(subset(sig_res, log2FoldChange > 0))\n",
    "    downreg <- as.data.frame(subset(sig_res, log2FoldChange < 0))\n",
    "    res <- as.data.frame(res)\n",
    "    \n",
    "    if (length(t(upreg)) > 0){\n",
    "        upreg$comparison <- title\n",
    "        upreg$gene_id <- rownames(upreg)\n",
    "    } #else{\n",
    "#         print(paste(\"No. upreg peaks with padj < 0.05 & fold change > 2 in comparison:\", title))\n",
    "#     }\n",
    "    \n",
    "    if (length(t(downreg)) > 0){\n",
    "        downreg$comparison <- title\n",
    "        downreg$gene_id <- rownames(downreg)\n",
    "    } #else{\n",
    "#         print(paste(\"No. downreg peaks with padj < 0.05 & fold change > 2 in comparison:\", title))\n",
    "#     }\n",
    "    res$comparison <- title\n",
    "    res$gene_id <- rownames(res)\n",
    "    \n",
    "    if (n == 1) {\n",
    "            de_genes <- rbind(upreg, downreg)\n",
    "            all_genes <- res\n",
    "    } else{\n",
    "        df <- rbind(upreg, downreg)\n",
    "        de_genes <- rbind(de_genes, df)\n",
    "        all_genes <- rbind(all_genes, res)\n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def upload2csvdb(df, gene_info, db, title):\n",
    "    '''Upload DESeq2 results to csvdb.\n",
    "       Seperate tables for signficant and all results'''\n",
    "    \n",
    "    df = pd.merge(df, gene_info[[\"gene_name\", \"gene_id\"]], how=\"inner\", on=\"gene_id\")\n",
    "    connect = sqlite3.connect(db)\n",
    "    \n",
    "    df.to_sql(title, connect, if_exists=\"replace\", index=False)\n",
    "    \n",
    "upload2csvdb(de_genes, gene_info, db, \"DESeq2_sig_results\")\n",
    "upload2csvdb(all_genes, gene_info, db, \"DESeq2_all_results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o go_comps\n",
    "\n",
    "# iterate over comparisons & get significant DE genes\n",
    "n = 0\n",
    "for (line in 1:nrow(condition_pairs)){\n",
    "    n = n + 1\n",
    "    sample <- condition_pairs[n, ]\n",
    "    a <- as.character(sample$a)\n",
    "    b <- as.character(sample$b)\n",
    "    title <- as.character(sample$comparison)\n",
    "\n",
    "    df <- as.data.frame(results(dds, contrast=c(\"category\", a, b)))\n",
    "\n",
    "    # get de genes, annotate and merge\n",
    "    up <- subset(df, padj < 0.05 & log2FoldChange > 1)\n",
    "    up$direction <- \"up\"\n",
    "    down <- subset(df, padj < 0.05 & log2FoldChange < -1)\n",
    "    down$direction <- \"down\"\n",
    "    \n",
    "    de_genes <- nrow(up) + nrow(down) # no de genes\n",
    "    \n",
    "    if (de_genes > 200){ # make sure there are enough de genes for GO analysis\n",
    "        if (n == 1){\n",
    "            goi <- rbind(up, down)\n",
    "            goi$category <- title\n",
    "        } else {\n",
    "            df2 <- rbind(up, down)\n",
    "            df2$category <- title\n",
    "            goi <- rbind(goi, df2)\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "# prepare gene lists for GO analysis\n",
    "goi$gene_id <- rownames(goi)\n",
    "goi$cluster <- paste0(goi$category, \"_\", goi$direction)\n",
    "goi <- goi[, c(\"gene_id\", \"cluster\")]\n",
    "\n",
    "# run GO analysis\n",
    "go_comps <- go_analysis(goi, bg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_comps[\"direction\"] = go_comps[\"cluster\"].apply(lambda x: x.split(\"_\")[-1])\n",
    "go_comps[\"comparison\"] = go_comps[\"cluster\"].apply(lambda x: '_'.join(x.split(\"_\")[:-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -h 800 -w 1200 -i go_comps\n",
    "\n",
    "# saveRDS(go_comps, file=\"DESeq2_comparisons_GO.rds\")\n",
    "# go_comps <- readRDS(\"DESeq2_comparisons_GO.rds\")\n",
    "\n",
    "plot_go <- function(df, pval=0.1, top=10, title=\"\"){\n",
    "    \n",
    "    df$Fold <- df$Significant / df$Expected\n",
    "\n",
    "    df <- subset(df, padj < pval)\n",
    "    \n",
    "    n <- 0\n",
    "    for (d in unique(df$direction)){\n",
    "        n <- n +1\n",
    "        cat <- subset(df, direction == d)\n",
    "        cat <- cat[order(cat$padj, decreasing=FALSE),  ]\n",
    "        cat <- head(cat, top)\n",
    "        \n",
    "        if (n==1){\n",
    "            res <- cat\n",
    "        } else {\n",
    "            res <- rbind(res, cat)\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    res <- res[order(res$cluster), ]\n",
    "    \n",
    "    p1 <- ggplot(res, aes(y=Fold, x=Term, fill=direction, alpha=-log10(padj))) + \n",
    "            geom_bar(stat=\"identity\", position=\"dodge\", colour=\"black\") + \n",
    "            geom_text(aes(label=Significant, y=Fold+0.2,), colour=\"black\", position=position_dodge(width=1), alpha=1, size=4) +\n",
    "            scale_x_discrete(limits=unique(res$Term), labels=str_wrap(unique(res$Term), 50)) +\n",
    "            scale_fill_manual(values=c(\"gray20\", \"red3\"), labels=c(\"Down\", \"Up\"), name=\"Direction\") +\n",
    "            scale_alpha(range=c(0.4, 1), name=\"-log10(FDR)\") +\n",
    "            coord_flip() +\n",
    "            labs(x=\"\", title=paste(\"Comparison:\", title)) +\n",
    "            theme_notebook() +\n",
    "            theme(legend.box=\"vertical\") +\n",
    "            geom_hline(yintercept=1, lty=\"dashed\")\n",
    "    \n",
    "    return(p1)\n",
    "}\n",
    "\n",
    "for (o in unique(go_comps$ontology)){\n",
    "    to_plot <- subset(go_comps, ontology==o)\n",
    "    for (c in unique(go_comps$comparison)){\n",
    "        p1 <- plot_go(subset(to_plot, comparison==c), top=10, title=c)\n",
    "        grid.arrange(top=textGrob(paste(\"Ontology:\", o), \n",
    "                                  gp=gpar(fontface=\"bold\", fontsize=22)), p1, ncol=1, nrow=1)\n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
