{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "<script>\n",
    "  jQuery(document).ready(function($) {\n",
    "\n",
    "  $(window).load(function(){\n",
    "    $('#preloader').fadeOut('slow',function(){$(this).remove();});\n",
    "  });\n",
    "\n",
    "  });\n",
    "</script>\n",
    "\n",
    "<style type=\"text/css\">\n",
    "  div#preloader { position: fixed;\n",
    "      left: 0;\n",
    "      top: 0;\n",
    "      z-index: 999;\n",
    "      width: 100%;\n",
    "      height: 100%;\n",
    "      overflow: visible;\n",
    "      background: #fff url('http://preloaders.net/preloaders/720/Moving%20line.gif') no-repeat center center;\n",
    "  }\n",
    "\n",
    "</style>\n",
    "\n",
    "<div id=\"preloader\"></div>"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "<script>\n",
    "  function code_toggle() {\n",
    "    if (code_shown){\n",
    "      $('div.input').hide('500');\n",
    "      $('#toggleButton').val('Show Code')\n",
    "    } else {\n",
    "      $('div.input').show('500');\n",
    "      $('#toggleButton').val('Hide Code')\n",
    "    }\n",
    "    code_shown = !code_shown\n",
    "  }\n",
    "\n",
    "  $( document ).ready(function(){\n",
    "    code_shown=false;\n",
    "    $('div.input').hide()\n",
    "  });\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" id=\"toggleButton\" value=\"Show Code\"></form>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load rpy2 magic\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "# to switch off warning messages\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# make default cell width 85% of available screen\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:85% !important; }</style>\"))\n",
    "\n",
    "# load R libraries & functions\n",
    "%R options(warn=-1)\n",
    "%R library(ggplot2)\n",
    "%R library(gplots)\n",
    "%R library(gridExtra)\n",
    "%R library(scales)\n",
    "%R library(ggrepel)\n",
    "%R library(wesanderson)\n",
    "%R library(reshape2)\n",
    "%R library(dplyr)\n",
    "%R library(plyr)\n",
    "%R library(grid)\n",
    "\n",
    "# load python modules\n",
    "import glob\n",
    "import re\n",
    "import sys\n",
    "import os\n",
    "import sqlite3\n",
    "import yaml\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "db = \"./csvdb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get options pipeline.yml\n",
    "with open(\"pipeline.yml\") as o:\n",
    "    opts = yaml.load(o)\n",
    "    \n",
    "anndb = opts[\"annotations\"][\"database\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_DataFrame(query, dbhandle=\"csvdb\"):\n",
    "    '''Fetch query results and returns them as a pandas dataframe'''\n",
    "\n",
    "    dbhandle = sqlite3.connect(dbhandle)\n",
    "\n",
    "    cc = dbhandle.cursor()\n",
    "    sqlresult = cc.execute(query).fetchall()\n",
    "    cc.close()\n",
    "\n",
    "    # see http://pandas.pydata.org/pandas-docs/dev/generated/\n",
    "    # pandas.DataFrame.from_records.html#pandas.DataFrame.from_records\n",
    "    # this method is design to handle sql_records with proper type\n",
    "    # conversion\n",
    "\n",
    "    field_names = [d[0] for d in cc.description]\n",
    "    pandas_DataFrame = pd.DataFrame.from_records(\n",
    "        sqlresult,\n",
    "        columns=field_names\n",
    "    )\n",
    "    return pandas_DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "# R functions\n",
    "theme_notebook <- function(base_size=18, base_family=\"helvetica\") {\n",
    "                  (theme_set(theme_minimal(base_size=18))\n",
    "                  + theme(plot.title = element_text(face=\"bold\", size=20, hjust=0.5),\n",
    "                             text = element_text(),\n",
    "                             axis.title = element_text(face=\"bold\",size = rel(1)),\n",
    "                             axis.title.y = element_text(angle=90,vjust=2, size=20),\n",
    "                             axis.title.x = element_text(vjust=-0.2, size=20),\n",
    "                             axis.text = element_text(size=20),\n",
    "                             axis.line = element_line(colour=\"black\"),\n",
    "                             axis.ticks = element_line(),\n",
    "                             legend.key = element_rect(colour = NA),\n",
    "                             legend.key.size= unit(0.5, \"cm\"),\n",
    "                             legend.margin = unit(0.5, \"cm\"),\n",
    "                             legend.text = element_text(size=14),\n",
    "                             legend.title = element_text(size=16),\n",
    "                             ))\n",
    "}\n",
    "\n",
    "# Set ggplot theme\n",
    "theme_set(theme_notebook(base_size=18))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Mapping QC\n",
    "***\n",
    "\n",
    "#### Sample information table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info = fetch_DataFrame('''select * from sample_info''', db)\n",
    "sample_info.drop(\"index\", axis=1, inplace=True)\n",
    "sample_info.head(len(sample_info))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i sample_info -o Palette\n",
    "\n",
    "# get pretty colours\n",
    "Palette <- wes_palette(sample(names(wes_palettes), 1))\n",
    "\n",
    "if (length(Palette) < length(unique(sample_info$category))){\n",
    "    while (length(Palette) < length(unique(sample_info$category))){\n",
    "        pal <- wes_palette(sample(names(wes_palettes), 1))\n",
    "        Palette <- unique(c(Palette, pal))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<br>\n",
    "\n",
    "\n",
    "### Picard alignment summary stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene_ids & gene_names for protein coding genes\n",
    "gene_info = fetch_DataFrame('''select distinct b.contig, b.start, b.end, a.gene_id, \n",
    "            a.gene_name from gene_info a, geneset_all_gtf_genome_coordinates \n",
    "            b where a.gene_id = b.gene_id and transcript_biotype = \"protein_coding\"''', anndb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "map_stats = fetch_DataFrame('''select PCT_ADAPTER, PCT_PF_READS_ALIGNED, READS_ALIGNED_IN_PAIRS, \n",
    "                                  TOTAL_READS, CATEGORY, sample_id from picardAlignmentSummary \n",
    "                                  where CATEGORY = \"PAIR\" ''', db)\n",
    "\n",
    "map_stats[\"sample_id\"] = map_stats[\"sample_id\"].apply(lambda x: os.path.basename(x))\n",
    "map_stats = pd.merge(map_stats, sample_info, how=\"inner\", on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i map_stats -w 1200 -h 700\n",
    "\n",
    "get_legend <- function(a.gplot){ \n",
    "  tmp <- ggplot_gtable(ggplot_build(a.gplot)) \n",
    "  leg <- which(sapply(tmp$grobs, function(x) x$name) == \"guide-box\") \n",
    "  legend <- tmp$grobs[[leg]] \n",
    "  return(legend)} \n",
    "                      \n",
    "total_reads <- ggplot(map_stats, aes(y=TOTAL_READS/2, x=sample_id, fill=condition, shape=treatment)) + \n",
    "                geom_point(size=6) + \n",
    "                scale_y_continuous(limits=c(0, max(map_stats$TOTAL_READS)/2)) +\n",
    "                geom_text_repel(aes(label=replicate), colour=\"black\") +\n",
    "                theme_notebook() +\n",
    "                scale_shape_manual(values=c(21,22,23,24)) +\n",
    "                scale_fill_manual(values=Palette) +\n",
    "                scale_colour_manual(values=rev(c(\"red\", \"darkgray\"))) +\n",
    "                guides(fill=guide_legend(override.aes=list(shape=21))) +\n",
    "                theme(axis.text.x=element_blank()) + labs(x=\"\", y=\"Total Read Pairs\")\n",
    "\n",
    "mapped <- ggplot(map_stats, aes(y=READS_ALIGNED_IN_PAIRS/2, x=sample_id, fill=condition, shape=treatment)) + \n",
    "                geom_point(size=6) + \n",
    "                scale_y_continuous(limits=c(0, max(map_stats$READS_ALIGNED_IN_PAIRS)/2)) +\n",
    "                geom_text_repel(aes(label=replicate), colour=\"black\") +\n",
    "                theme_notebook() +\n",
    "                scale_shape_manual(values=c(21,22,23,24)) +\n",
    "                scale_fill_manual(values=Palette) +\n",
    "                scale_colour_manual(values=rev(c(\"red\", \"darkgray\"))) +\n",
    "                guides(fill=guide_legend(override.aes=list(shape=21))) +\n",
    "                theme(axis.text.x=element_blank()) + labs(x=\"\", y=\"Mapped Pairs\")\n",
    "\n",
    "\n",
    "pct_mapped <- ggplot(map_stats, aes(y=PCT_PF_READS_ALIGNED*100, x=sample_id, fill=condition, shape=treatment)) + \n",
    "                geom_point(size=6) + \n",
    "                scale_y_continuous(limits=c(0, 110)) +\n",
    "                geom_text_repel(aes(label=replicate), colour=\"black\") +\n",
    "                theme_notebook() +\n",
    "                scale_shape_manual(values=c(21,22,23,24)) +\n",
    "                scale_fill_manual(values=Palette) +\n",
    "                scale_colour_manual(values=rev(c(\"red\", \"darkgray\"))) +\n",
    "                guides(fill=guide_legend(override.aes=list(shape=21))) +\n",
    "                theme(axis.text.x=element_blank()) + labs(x=\"\", y=\"% Mapped in Pairs\")\n",
    "\n",
    "\n",
    "pct_adaptor <- ggplot(map_stats, aes(y=PCT_ADAPTER, x=sample_id,  fill=condition, shape=treatment)) + \n",
    "                geom_point(size=6) + \n",
    "                geom_text_repel(aes(label=replicate), colour=\"black\") +\n",
    "                theme_notebook() +\n",
    "                scale_shape_manual(values=c(21,22,23,24)) +\n",
    "                scale_fill_manual(values=Palette) +\n",
    "                scale_colour_manual(values=rev(c(\"red\", \"darkgray\"))) +\n",
    "                guides(fill=guide_legend(override.aes=list(shape=21))) +\n",
    "                theme(axis.text.x=element_blank()) + labs(x=\"\", y=\"% adaptor\")\n",
    "\n",
    "                      \n",
    "legend <- get_legend(total_reads + theme(legend.direction=\"horizontal\"))\n",
    "\n",
    "grid.arrange(top=textGrob(\"Mapping Stats\", gp=gpar(fontfamily=\"Helvetica\", fontface=\"bold\", fontsize=23)), \n",
    "             total_reads + theme(legend.position=\"none\"), mapped + theme(legend.position=\"none\"), \n",
    "             pct_mapped + theme(legend.position=\"none\"), pct_adaptor + theme(legend.position=\"none\"), \n",
    "             ncol=2, nrow=2, bottom=legend)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<br>\n",
    "\n",
    "\n",
    "### Picard RNA-seq metrics\n",
    "\n",
    "#### Mapping annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_map_stats_pct = fetch_DataFrame('''select PCT_USABLE_BASES, PCT_RIBOSOMAL_BASES, PCT_CODING_BASES, \n",
    "                                          PCT_UTR_BASES, PCT_INTRONIC_BASES, PCT_INTERGENIC_BASES, sample_id\n",
    "                                          from picardRNAseqMetrics ''', db)\n",
    "\n",
    "rna_map_stats_pct[\"PCT_RIBOSOMAL_BASES\"] = rna_map_stats_pct[\"PCT_RIBOSOMAL_BASES\"].apply(lambda x: float(0) if str(x) == \"None\" else x)\n",
    "rna_map_stats_pct[\"sample_id\"] = rna_map_stats_pct[\"sample_id\"].apply(lambda x: os.path.basename(x))\n",
    "rna_map_stats_pct = pd.merge(rna_map_stats_pct, sample_info, how=\"inner\", on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i rna_map_stats_pct -w 1200 \n",
    "\n",
    "# melt <- melt(rna_map_stats, id_var=c(\"sample_id\", \"TF\", \"Day\", \"replicate\", \"LPS\"))\n",
    "\n",
    "# p1 <- ggplot(melt, aes(y=value, x=sample_id, fill=variable)) +\n",
    "#            geom_bar(stat=\"identity\",colour=\"black\") +\n",
    "#            scale_fill_manual(values=Palette) +\n",
    "#            coord_flip() #+\n",
    "# #            facet_wrap(~ facet)\n",
    "\n",
    "melt2 <- melt(rna_map_stats_pct, id_var=c(\"sample_id\", \"condition\", \"treatment\", \"replicate\"))\n",
    "\n",
    "melt2$facet <- \"feature / PF bases\"\n",
    "melt2$facet[melt2$variable==\"PCT_USABLE_BASES\"] <- \"useable (mRNA / PF bases)\"\n",
    "\n",
    "p2 <- ggplot(melt2, aes(y=value, x=sample_id, fill=variable)) +\n",
    "           geom_bar(stat=\"identity\",colour=\"black\") +\n",
    "           theme_notebook() +\n",
    "           scale_fill_manual(values=c(Palette, \"darkseagreen3\")) +\n",
    "           coord_flip() +\n",
    "           facet_wrap(~ facet) +\n",
    "           labs(y=\"\", x=\"\") +\n",
    "           scale_y_continuous(labels=percent)\n",
    "        \n",
    "p2\n",
    "\n",
    "# Illumina sequencers perform an internal quality filtering procedure called chastity filter, \n",
    "# and reads that pass this filter are called PF for pass-filter. \n",
    "# According to Illumina, chastity is defined as the ratio of the brightest base intensity \n",
    "# divided by the sum of the brightest and second brightest base intensities. \n",
    "# Clusters of reads pass the filter if no more than 1 base call has a chastity value \n",
    "# below 0.6 in the first 25 cycles. This filtration process removes the least reliable \n",
    "# clusters from the image analysis results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PF bases = pass-filter (Illumina internal quality filtering), only reads passing filter are counted\n",
    "\n",
    "<br>\n",
    "\n",
    "#### Strand orientation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_map_strand = fetch_DataFrame('''select CORRECT_STRAND_READS, INCORRECT_STRAND_READS, \n",
    "                                  NUM_R1_TRANSCRIPT_STRAND_READS, NUM_R2_TRANSCRIPT_STRAND_READS, \n",
    "                                  sample_id from picardRNAseqMetrics ''', db)\n",
    "\n",
    "rna_map_strand[\"sample_id\"] = rna_map_strand[\"sample_id\"].apply(lambda x: os.path.basename(x))\n",
    "rna_map_strand = pd.merge(rna_map_strand, sample_info, how=\"inner\", on=\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i rna_map_strand -w 1200\n",
    "\n",
    "melt <- melt(rna_map_strand, id_vars=c(\"sample_id\", \"condtion\", \"treatment\", \"replicate\"))\n",
    "\n",
    "p1 <- ggplot(melt, aes(y=value, x=sample_id, fill=variable)) + \n",
    "        geom_bar(stat=\"identity\", colour=\"black\") +\n",
    "        scale_fill_manual(values=Palette) +\n",
    "        theme_notebook() +\n",
    "        coord_flip() +\n",
    "        labs(y=\"\", x=\"\")\n",
    "        \n",
    "p1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- R1_TRANSCRIPT_STRAND_READS\n",
    "    - The fraction of reads that support the model where R1 is on the strand of transcription and R2 is on the opposite strand. \n",
    "    - For unpaired reads, it is the fraction of reads that are on the transcription strand (out of all the reads).\n",
    "- R2_TRANSCRIPT_STRAND_READS\n",
    "    - The fraction of reads that support the model where R2 is on the strand of transcription and R1 is on the opposite strand. \n",
    "    - For unpaired reads, it is the fraction of reads that are on opposite strand than that of the the transcription strand (out of all the reads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_map_bias = fetch_DataFrame('''select MEDIAN_3PRIME_BIAS, MEDIAN_5PRIME_BIAS, \n",
    "                                  MEDIAN_5PRIME_TO_3PRIME_BIAS, MEDIAN_CV_COVERAGE,\n",
    "                                  sample_id from picardRNAseqMetrics''', db)\n",
    "\n",
    "rna_map_bias[\"sample_id\"] = rna_map_bias[\"sample_id\"].apply(lambda x: os.path.basename(x))\n",
    "rna_map_bias = pd.merge(rna_map_bias, sample_info, how=\"inner\", on=\"sample_id\")\n",
    "# rna_map_bias.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i rna_map_bias -w 1200 -h 1800\n",
    "\n",
    "melt <- melt(rna_map_bias[, c(\"MEDIAN_3PRIME_BIAS\", \"MEDIAN_5PRIME_BIAS\", \"sample_id\", \"condition\", \"treatment\", \"replicate\")],\n",
    "             id_vars=c(\"sample_id\", \"condition\", \"treatment\", \"replicate\"))\n",
    "\n",
    "\n",
    "p1 <- ggplot(rna_map_bias, aes(y=MEDIAN_CV_COVERAGE, x=sample_id, fill=condition)) +\n",
    "        geom_bar(stat=\"identity\", colour=\"black\") +\n",
    "        scale_fill_manual(values=Palette) +\n",
    "        theme_notebook() +\n",
    "        coord_flip() +\n",
    "        labs(x=\"\", caption=\"Ideal value = 0.\\n\",\n",
    "            title=\"The median coefficient of variation (stdev/mean) for coverage values of the top 1,000 genes\")\n",
    "        \n",
    "        \n",
    "p2 <- ggplot(melt, aes(y=value, x=sample_id, fill=variable)) +\n",
    "        geom_bar(stat=\"identity\", colour=\"black\") +\n",
    "        scale_fill_manual(values=Palette) +\n",
    "        theme_notebook() +\n",
    "        coord_flip() +\n",
    "        scale_y_continuous(labels=percent) +\n",
    "        labs(y=\"\", x=\"\", title=\"The median 3 and 5 prime bias of the 1,000 most highly expressed transcripts\",\n",
    "            caption=\"\\n\")\n",
    "        \n",
    "\n",
    "p3 <- ggplot(rna_map_bias, aes(y=MEDIAN_5PRIME_TO_3PRIME_BIAS, x=sample_id, fill=condition)) +\n",
    "        geom_bar(stat=\"identity\", colour=\"black\") +\n",
    "        scale_fill_manual(values=Palette) +\n",
    "        theme_notebook() +\n",
    "        coord_flip() +\n",
    "        labs(x=\"\", caption=\"\\n\",\n",
    "             title=\"The ratio of coverage at the 5 prime end to the 3 prime end based on the \\n1,000 most highly expressed transcripts.\")\n",
    "        \n",
    "        \n",
    "grid.arrange(p1, p2, p3, ncol=1, nrow=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<br>\n",
    "\n",
    "### Salmon TPMs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Transcripts per million (TPM) were quantitated using Salmon and upper quantile normalised for between sample comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "salmon_genes = fetch_DataFrame('''select * from salmon_genes''', db) # get data\n",
    "\n",
    "# normalise to upper quantiles for between sample comparison\n",
    "salmon_genes.index = salmon_genes[\"gene_id\"]\n",
    "salmon_genes.drop(\"gene_id\", inplace=True, axis=1)\n",
    "salmon_genes = salmon_genes.div(salmon_genes.quantile(q=0.75, axis=0), axis=1)\n",
    "salmon_genes.reset_index(inplace=True)\n",
    "\n",
    "# subset on protein coding genes\n",
    "salmon_genes = pd.merge(salmon_genes, gene_info[[\"gene_id\"]], how=\"inner\", on=\"gene_id\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "### Number of genes detected per sample "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many genes are expressed across samples\n",
    "\n",
    "def count_genes(df, sample_info, threshold):\n",
    "    \n",
    "    df = df.drop(\"gene_id\", axis=1)\n",
    "    \n",
    "    expressed_genes = {}\n",
    "    n=0\n",
    "    for column in df:\n",
    "        \n",
    "        n=n+1\n",
    "        if n == 1:\n",
    "            tpm_count = pd.DataFrame(df.loc[df[column] > threshold][[column]].count(numeric_only=True))\n",
    "        else:\n",
    "            df2 = pd.DataFrame(df.loc[df[column] > threshold][[column]].count(numeric_only=True))\n",
    "            tpm_count = tpm_count.append(df2)\n",
    "\n",
    "    tpm_count[\"sample_id\"] = tpm_count.index.values\n",
    "    tpm_count.columns = [\"no_genes\", \"sample_id\"]\n",
    "    tpm_count = pd.merge(tpm_count, sample_info, how=\"inner\", on=\"sample_id\")\n",
    "    \n",
    "    return tpm_count\n",
    "\n",
    "\n",
    "tpm_count_0 = count_genes(salmon_genes, sample_info, 0)\n",
    "tpm_count_1 = count_genes(salmon_genes, sample_info, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i tpm_count_0,tpm_count_1 -w 1200 -h 400\n",
    "\n",
    "no_genes_0 <- ggplot(tpm_count_0, aes(y=no_genes, x=sample_id, fill=condition, shape=treatment)) + \n",
    "                geom_point(size=6) + scale_y_continuous(limits=c(0, 22000)) +\n",
    "                geom_text(aes(label=replicate), colour=\"black\", nudge_y=1800) +\n",
    "                theme_notebook() +\n",
    "                scale_shape_manual(values=c(21,22,23,24)) +\n",
    "                scale_alpha_discrete(range=c(0.6,1)) +\n",
    "                scale_fill_manual(values=Palette) +\n",
    "                scale_colour_manual(values=c(\"black\", \"red3\")) +\n",
    "                theme(axis.text.x=element_blank()) + labs(x=\"\", y=\"No. Genes (TPM >0)\") +\n",
    "                guides(fill=guide_legend(override.aes=list(shape=21))) +\n",
    "                theme(legend.box=\"right\")\n",
    "\n",
    "no_genes_1 <- ggplot(tpm_count_1, aes(y=no_genes, x=sample_id, fill=condition, shape=treatment)) + \n",
    "                geom_point(size=6) + scale_y_continuous(limits=c(0, 22000)) +\n",
    "                geom_text(aes(label=replicate), colour=\"black\", nudge_y=1800) +\n",
    "                theme_notebook() +\n",
    "                scale_alpha_discrete(range=c(0.6,1)) +\n",
    "                scale_shape_manual(values=c(21,22,23,24)) +\n",
    "                scale_fill_manual(values=Palette) +\n",
    "                scale_colour_manual(values=c(\"black\", \"red3\")) +\n",
    "                theme(axis.text.x=element_blank()) + labs(x=\"\", y=\"No. Genes (TPM >1)\") +\n",
    "                guides(fill=guide_legend(override.aes=list(shape=21))) +\n",
    "                theme(legend.box=\"right\")\n",
    "\n",
    "grid.arrange(no_genes_0, no_genes_1, ncol=2, nrow=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "### Pearson correlation\n",
    "- based on Manhattan distances between samples\n",
    "\n",
    "#### For all detected genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i salmon_genes -w 750 -h 650\n",
    "\n",
    "# pearson correlation on all gene TPMs \n",
    "\n",
    "pal <- colorRampPalette(c(\"#3B9AB2\", \"#EBCC2A\", \"#F21A00\"))(300)\n",
    "\n",
    "plot_cor <- function(df) {\n",
    "    \n",
    "    df$gene_id <- NULL\n",
    "    cm <- data.matrix(log2(df +1))\n",
    "    m <- cor(cm, method=\"pearson\", use=\"all\")\n",
    "    \n",
    "    heatmap.2(m, \n",
    "      trace=\"none\",\n",
    "      key.xlab = \"Pearson Correlation\",\n",
    "      key.ylab = \"\",\n",
    "      col=pal,\n",
    "      cexRow=1.5,\n",
    "      cexCol=1.5,\n",
    "      density.info=c(\"none\"),\n",
    "      mar=c(20,20),\n",
    "      key.title=NA,\n",
    "      lhei=c(0.3,2.2),\n",
    "      lwid=c(0.6,2),\n",
    "      key.par=list(cex.lab=2, cex.axis=1.5))}    \n",
    "\n",
    "plot_cor(salmon_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "#### For top 5,000 genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "salmon_genes_top = salmon_genes.copy(deep=True)\n",
    "salmon_genes_top[\"mean\"] = salmon_genes_top.mean(axis=1)\n",
    "salmon_genes_top = salmon_genes_top.sort_values(\"mean\", ascending=False).head(5000)\n",
    "salmon_genes_top.drop(\"mean\", inplace=True, axis=1)\n",
    "\n",
    "# salmon_genes_top.head()\n",
    "# print(len(salmon_genes_top))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i salmon_genes_top -w 750 -h 650\n",
    "\n",
    "# pearson correlation on top 5000 gene TPMs  \n",
    "\n",
    "plot_cor(salmon_genes_top)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<br>\n",
    "\n",
    "### Replicate correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use sample information to get no. replicates & conditions\n",
    "rep_pairs = sample_info.copy(deep=True)\n",
    "rep_pairs = rep_pairs.pivot(\"category\", \"replicate\", \"sample_id\").transpose()\n",
    "rep_pairs.columns.name = None\n",
    "rep_pairs.index.name = None\n",
    "\n",
    "# report replicates to dict\n",
    "reps = {}\n",
    "for col in rep_pairs.columns:\n",
    "    reps[col]=[rep_pairs[col].iloc[0], rep_pairs[col].iloc[1]]\n",
    "\n",
    "# print(reps)\n",
    "\n",
    "# log2 transfrom Salmon TPMs\n",
    "salmon_genes.set_index(\"gene_id\", inplace=True)\n",
    "salmon_genes.fillna(0, inplace=True)\n",
    "counts = np.log2(salmon_genes +1)\n",
    "\n",
    "# get palette\n",
    "pal = Palette\n",
    "if len(rep_pairs.transpose()) > len(pal):\n",
    "    import random\n",
    "    \n",
    "    extra_colours = sns.xkcd_rgb.keys() # all 954 colours\n",
    "    pal2 = sns.xkcd_palette(extra_colours)\n",
    "    pal = pal + random.sample(pal2, len(pal2))\n",
    "    \n",
    "sns.set(style=\"whitegrid\", palette=\"muted\")# set seaborn theme\n",
    "\n",
    "# ignore seaborn deprication warning\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# use dict to subset df of normalised counts & plot rep correlations\n",
    "n = 0\n",
    "for key in reps:\n",
    "    n = n + 1\n",
    "    c = n - 1\n",
    "    df = counts[reps[key]]\n",
    "    df.columns = [\"Rep1\", \"Rep2\"]\n",
    "    p = sns.jointplot(data=df, y=\"Rep1\", x=\"Rep2\", kind=\"reg\", height=7, color=pal[c])\n",
    "    plt.subplots_adjust(top=0.9)\n",
    "    p.fig.suptitle(key) # add title\n",
    "    p.annotate(stats.pearsonr)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
